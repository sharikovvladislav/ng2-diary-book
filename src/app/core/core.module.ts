import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import {
  MaterialModule,
  MdButtonModule,
  MdIconModule,
} from '@angular/material';
import { RouterModule } from '@angular/router';

import { HTTP_INTERCEPTORS, HttpClientModule } from '@angular/common/http';
import { AuthInterceptor } from './modules/http-interceptors/auth-append-token';

import { AppComponent } from './containers/app';
import { NotFoundPageComponent } from './containers/not-found-page';
import { LayoutComponent } from './components/layout';
import { NavItemComponent } from './components/nav-item';
import { SidenavComponent } from './components/sidenav';
import { ToolbarComponent } from './components/toolbar';
import { NotLoggedInComponent } from './components/not-logged-in';
import { CommonShowIfLoggedInComponent } from './containers/common-show-if-logged-in';
import { LoaderComponent } from './containers/loader';
import { BreadcrumbsComponent } from './components/breadcrumbs';
import { BreadcrumbsContainer } from './containers/breadcrumbs';

import { BreadcrumbsService } from './services/breadcrumbs';

import { PipesModule } from '../shared/pipes';
import { EffectsModule } from '@ngrx/effects';
import { RouterEffects } from './effects/router';
import { FormsModule } from '@angular/forms';
import { TagsAutoCompleteModule } from './features/components/tags-autocomplete/tags-autocomplete.module';

export const COMPONENTS = [
  AppComponent,
  NotFoundPageComponent,
  LayoutComponent,
  NavItemComponent,
  SidenavComponent,
  ToolbarComponent,
  NotLoggedInComponent,
  CommonShowIfLoggedInComponent,
  LoaderComponent,
  BreadcrumbsComponent,
  BreadcrumbsContainer,
];

const MD_MODULES = [MaterialModule, MdButtonModule, MdIconModule];

@NgModule({
  imports: [
    CommonModule,
    FormsModule,
    ...MD_MODULES,
    RouterModule,
    PipesModule,
    HttpClientModule,
    EffectsModule.forRoot([RouterEffects]),
    TagsAutoCompleteModule,
  ],
  declarations: COMPONENTS,
  exports: [
    ...COMPONENTS,
    CommonModule,
    FormsModule,
    ...MD_MODULES,
    HttpClientModule,
  ],
  providers: [
    BreadcrumbsService,
    {
      provide: HTTP_INTERCEPTORS,
      useClass: AuthInterceptor,
      multi: true,
    },
  ],
})
export class CoreModule {
  static forRoot() {
    return {
      ngModule: CoreModule,
    };
  }
}
