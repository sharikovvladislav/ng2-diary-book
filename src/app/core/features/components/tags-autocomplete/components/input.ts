import {
  Component,
  ElementRef,
  EventEmitter,
  Input,
  OnInit,
  Output,
  ViewChild,
} from '@angular/core';
import { Tag } from 'ng2-diary-book-shared-models';

@Component({
  selector: 'tags-internal-auto-complete-input',
  template: `
    <div class="container">
      <div class="selectedTags">
        <span *ngFor="let tag of selectedTags" class="tag">
          {{tag.name}}&nbsp;<span (click)="deleteTag.emit(tag);">X</span>
        </span>
        <div class="input">
          <input #input
                 placeholder="{{placeholder}}"
                 (keydown.backspace)="onBackspacePressed(input.value)"
                 [(ngModel)]="inputValue"
                 (ngModelChange)="inputChanged.emit(input.value);"
          />
          <tags-auto-complete-loader class="loader" [class.hidden]="!showLoader"></tags-auto-complete-loader>
        </div>
      </div>
    </div>
  `,
  styleUrls: ['./input.styles.css'],
})
export class TagsAutoCompleteInputComponent implements OnInit {
  @ViewChild('input') inputRef: ElementRef;

  @Input() selectedTags: Tag[] = [];
  @Input() placeholder = '';
  @Input() showLoader = false;

  inputValue = '';

  @Output() deleteTag = new EventEmitter<Tag>();
  @Output() inputChanged = new EventEmitter<string>();

  clearInput(): void {
    this.inputValue = '';
  }

  focusInput(): void {
    this.inputRef.nativeElement.focus();
  }

  onBackspacePressed({ length: inputValueLength }) {
    if (inputValueLength === 0) {
      this.deleteTag.emit(this.selectedTags[this.selectedTags.length - 1]);
    }
  }

  ngOnInit() {}
}
