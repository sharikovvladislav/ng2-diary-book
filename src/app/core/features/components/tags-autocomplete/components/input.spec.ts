import { MyTestBed } from '../../../../../test/helpers/my-test-bed';

import { TagsAutoCompleteInputComponent } from './input';
import { FormsModule } from '@angular/forms';
import { TagsAutoCompleteLoaderComponent } from './loader';
import { ComponentFixture, fakeAsync, tick } from '@angular/core/testing';
import { DebugElement, EventEmitter } from '@angular/core';
import { By } from '@angular/platform-browser';

describe('TagsAutoCompleteInputComponent', () => {
  let component: TagsAutoCompleteInputComponent;
  let fixture: ComponentFixture<TagsAutoCompleteInputComponent>;
  let de: DebugElement;

  describe('tests by snapshots', () => {
    beforeEach(async () => {
      MyTestBed.configureTestingModule({
        imports: [FormsModule],
        declarations: [
          TagsAutoCompleteInputComponent,
          TagsAutoCompleteLoaderComponent,
        ],
      })
        .shallowRenderComponent(TagsAutoCompleteLoaderComponent)
        .compileComponents();
    });

    beforeEach(() => {
      fixture = MyTestBed.createComponent(TagsAutoCompleteInputComponent);
      component = fixture.componentInstance;

      fixture.detectChanges();
    });

    it('initial rendering is ok', () => {
      expect(fixture).toMatchSnapshot();
    });
  });

  describe('tests in general', () => {
    beforeEach(async () => {
      MyTestBed.configureTestingModule({
        imports: [FormsModule],
        declarations: [
          TagsAutoCompleteInputComponent,
          TagsAutoCompleteLoaderComponent,
        ],
      }).compileComponents();
    });

    beforeEach(() => {
      fixture = MyTestBed.createComponent(TagsAutoCompleteInputComponent);
      component = fixture.componentInstance;
      de = fixture.debugElement;

      fixture.detectChanges();
    });

    it('placeholder input renders correctly', () => {
      component.placeholder = 'abc';

      fixture.detectChanges();

      const inputNode = de.query(By.css('input'));

      expect(inputNode.nativeElement.placeholder).toBe('abc');
    });

    it(
      'clear input if event emitter at input emitted some stuff',
      fakeAsync(() => {
        expect.assertions(2);

        component.inputValue = 'some value';

        fixture.detectChanges();

        tick();

        expect(de.query(By.css('div > input')).nativeElement.value).toBe(
          'some value',
        );

        component.clearInput();

        fixture.detectChanges();

        tick();

        expect(de.query(By.css('div > input')).nativeElement.value).toBe('');
      }),
    );

    describe('backspace checks', () => {
      it('when internal on backspace method called, should emit value to deleteTag output', () => {
        expect.assertions(1);

        component.selectedTags = [{ name: 'first' }, { name: 'second' }];

        const spy = jest.fn();

        component.deleteTag.subscribe(spy);

        component.onBackspacePressed({ length: 0 });

        expect(spy).toHaveBeenCalledWith(component.selectedTags[1]);
      });

      it('when backspace pressed, event handler is called', () => {
        const spy = jest.spyOn(component, 'onBackspacePressed');
        const event = new KeyboardEvent('keydown', {
          key: 'backspace',
        });
        de.query(By.css('input')).nativeElement.dispatchEvent(event);

        fixture.detectChanges();

        expect(spy).toHaveBeenCalled();

        spy.mockReset();
      });
    });

    it('focus on element is called when handler is called', () => {
      const input = de.query(By.css('input')).nativeElement;

      const spy = jest.spyOn(input, 'focus');

      component.focusInput();

      expect(spy).toHaveBeenCalled();
    });

    it('should emit deleted tag to output when clicked on delete button', () => {
      const spy = jest.fn();
      component.deleteTag.subscribe(spy);

      component.selectedTags = [{ name: 'A' }, { name: 'B' }];

      fixture.detectChanges();

      const selectedTags = de.queryAll(By.css('.selectedTags > .tag'));
      expect(selectedTags.length).toBe(2);

      const deleteButton = selectedTags[0].query(By.css('span'));
      deleteButton.nativeElement.dispatchEvent(new MouseEvent('click'));

      fixture.detectChanges();

      expect(spy).toHaveBeenCalledWith(component.selectedTags[0]);
    });

    it('should render selectedTags accordingly to input', () => {
      expect(de.queryAll(By.css('.selectedTags > .tag')).length).toBe(0);

      component.selectedTags = [{ name: '1' }, { name: '2' }];

      fixture.detectChanges();

      expect(de.queryAll(By.css('.selectedTags > .tag')).length).toBe(2);

      component.selectedTags = [{ name: '1' }, { name: '2' }, { name: 'x' }];

      fixture.detectChanges();

      expect(de.queryAll(By.css('.selectedTags > .tag')).length).toBe(3);
    });
  });
});
