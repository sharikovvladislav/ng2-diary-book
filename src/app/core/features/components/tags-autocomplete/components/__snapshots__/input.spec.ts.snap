// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagsAutoCompleteInputComponent tests by snapshots initial rendering is ok 1`] = `
<tags-internal-auto-complete-input
  deleteTag={[Function EventEmitter]}
  inputChanged={[Function EventEmitter]}
  inputRef={[Function ElementRef]}
  inputValue=""
  placeholder=""
  selectedTags={[Function Array]}
  showLoader="false"
>
  <div
    class="container"
  >
    <div
      class="selectedTags"
    >
      <div
        class="input"
      >
        <input
          ng-reflect-model=""
          placeholder=""
          class="ng-untouched ng-pristine ng-valid"
        />
        <tags-auto-complete-loader
          class="loader hidden"
        >
          tags-auto-complete-loader component stub
        </tags-auto-complete-loader>
      </div>
    </div>
  </div>
</tags-internal-auto-complete-input>
`;
